icu4c/source/common/serv.cpp

```diff
@@ -524,22 +519,26 @@ ICUService::getKey(ICUServiceKey& key, UnicodeString* actualReturn, const ICUSer
             // don't want to keep querying on an id that's going to
             // fallback to the one that succeeded, we want to hit the
             // cache the first time next goaround.
-            if (cacheDescriptorList._obj == NULL) {
-                cacheDescriptorList._obj = new UVector(uprv_deleteUObject, NULL, 5, status);
+            if (cacheDescriptorList.isNull()) {
+                cacheDescriptorList.adoptInsteadAndCheckErrorCode(new UVector(uprv_deleteUObject, NULL, 5, status), status);
                 if (U_FAILURE(status)) {
                     return NULL;
                 }
             }
-            UnicodeString* idToCache = new UnicodeString(currentDescriptor);
-            if (idToCache == NULL || idToCache->isBogus()) {
+
+            LocalPointer<UnicodeString> idToCache(new UnicodeString(currentDescriptor), status);
+            if (U_FAILURE(status)) {
+                return NULL;
+            }
+            if (idToCache->isBogus()) {
                 status = U_MEMORY_ALLOCATION_ERROR;
                 return NULL;
             }
-
-            cacheDescriptorList._obj->addElement(idToCache, status);
+            cacheDescriptorList->addElement(idToCache.getAlias(), status);
             if (U_FAILURE(status)) {
                 return NULL;
             }
+            idToCache.orphan(); // cacheDescriptorList now owns the string.
```

Comment:
Hmm. This is annoying and error-prone. Please submit a ticket for adding a UVector.adoptElement() function.

Reply:
Thanks, that's a good point. I filed a ticket here for it: https://unicode-org.atlassian.net/browse/ICU-21662